cupy/cuda/memory.pyx

```diff
@@ -1973,3 +1973,22 @@ cdef class PythonFunctionAllocator:
             size, self._malloc_func,
             self._free_func, device.get_device_id())
         return MemoryPointer(mem, 0)
+
+
+cdef dict _allocations = dict()
+
+
+cdef public void* cupy_malloc_ext(
+        ssize_t size, int device, void* stream) with gil:
+    # TODO(ecastill) dynamically change stream & device?
+    cdef curr_stream = stream_module.get_current_stream().ptr
+    if <intptr_t>stream != <intptr_t>curr_stream:
+        raise RuntimeError(f'stream must be the current one')
+    mem = alloc(size)
```

Comment:
I have two conflicting concerns here:
1. Shouldn't the (user-provided) `stream` be honored? Currently this assumes users manage CUDA streams using CuPy's machinery (ex: `with stream as s:`), which may or may not be feasible when an external library is hooked with CuPy's pool.
2. On the other hand, if `stream` is not the (CuPy) current stream, `alloc()` is incorrect because it places the chunk under the wrong stream, potentially breaking the stream ordering. If 1 is addressed, we'd need to do
```python
with ExternalStream(ptr=stream):
    mem = alloc(size)
```
instead. The current constraint (that `stream` is made current) makes the situation a bit easier.

Reply:
You are right :), I just sent an attempt to do it

Reply:
We originally thought of asking users to use [`ppe.cuda.stream()`](https://github.com/pfnet/pytorch-pfn-extras/blob/5c8be80b1f70719e501e01ef2b67f620aad3ad46/pytorch_pfn_extras/cuda/_allocator.py#L13) (which is an utility we use to switch both PyTorch/CuPy streams), but I agree that there is a case of hooking external libraries that cannot be modified. Let's honor the stream provided.